### PR TITLE
[Serialization] Improve extensions of nested types with the same name

### PR DESCRIPTION
**Pull request by jrose-apple**
Previously looking up an extension would result in all extensions for types with the same name (nested or not) being deserialized; this could even bring in base types that had not been deserialized yet. Add in a string to distinguish an extension's base type; in the top-level case this is just a module name, but for nested types it's a full mangled name.

This is a little heavier than I'd like it to be, since it means we mangle names and then throw them away, and since it means there's a whole bunch of extra string data in the module just for uniquely identifying a declaration. But it's correct, and does less work than before, and fixes a circularity issue with a nested type A.B.A that apparently used to work.

[SR-3915](https://bugs.swift.org/browse/SR-3915)

### CODE REVIEW

**Comment by slavapestov** [comment]
This looks like it needs a rebase? Excellent work though :)

**Comment by jrose-apple** [comment]
Yeah, it's "just" the module format version but I'm taking the opportunity to retest locally. Thanks!

**Comment by jrose-apple** [comment]
@swift-ci Please test

File: lib/Serialization/Serialization.h [hunk 1]
```diff
@@ -128,12 +128,16 @@ class Serializer {
 
   // In-memory representation of what will eventually be an on-disk
   // hash table of all defined Objective-C methods.
-  using ObjCMethodTable = llvm::DenseMap<ObjCSelector, ObjCMethodTableData>;
+  using ObjCMethodTable = llvm::MapVector<ObjCSelector, ObjCMethodTableData>;
```
Comment (DougGregor) [inline lib/Serialization/Serialization.h]: Good drive-by

**Review by DougGregor — APPROVED** [review]
It does feel a bit heavy to use the mangled name here, but it's definitely a good, descriptive key.

**Comment by jrose-apple** [comment]
@swift-ci Please smoke test
